src/lib.rs

```diff
@@ -306,17 +306,28 @@ impl<T: Contribution> From<WireMessageKind<T>> for WireMessage<T> {
     }
 }
 
+/// A serialized `WireMessage` signed by the sender.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedWireMessage {
+    message: Vec<u8>,
+    sig: Signature,
+}
+
 /// A stream/sink of `WireMessage`s connected to a socket.
 #[derive(Debug)]
 pub struct WireMessages<T> {
     framed: Framed<TcpStream, LengthDelimitedCodec>,
+    our_key: SecretKey,
+    their_key: Option<PublicKey>,
     _t: PhantomData<T>,
 }
 
 impl<T: Contribution> WireMessages<T> {
-    pub fn new(socket: TcpStream) -> WireMessages<T> {
+    pub fn new(socket: TcpStream, our_key: SecretKey) -> WireMessages<T> {
         WireMessages {
             framed: Framed::new(socket, LengthDelimitedCodec::new()),
+            our_key,
+            their_key: None,
```

Comment:
A `WireMessages` struct is shared and used by all communications to all nodes. Saving anything node-specific isn't appropriate, node specific info is stored in the `Peer` struct.